File: client/web/sudoo/lib/data/repository/auth_repository_impl.dart
import 'dart:convert';

import 'package:shared_preferences/shared_preferences.dart';
import 'package:sudoo/data/api/auth/auth_api_service.dart';
import 'package:sudoo/data/api/auth/request/account_request.dart';
import 'package:sudoo/data/api/auth/request/change_password_request.dart';
import 'package:sudoo/data/api/auth/request/verify_otp_request.dart';
import 'package:sudoo/data/api/handle_response.dart';
import 'package:sudoo/data/config/pref_keys.dart';
import 'package:sudoo/data/dto/auth/token_dto.dart';
import 'package:sudoo/domain/core/data_state.dart';
import 'package:sudoo/domain/model/auth/account.dart';
import 'package:sudoo/domain/model/auth/auth_config.dart';
import 'package:sudoo/domain/model/auth/verify_otp.dart';
import 'package:sudoo/domain/repository/auth_repository.dart';
import 'package:sudoo/extensions/string_ext.dart';

import '../../domain/common/Constants.dart';
import '../../domain/model/auth/token.dart';

class AuthRepositoryImpl with HandleResponse implements AuthRepository {
  final AuthApiService authApiService;
  final SharedPreferences pref;

  const AuthRepositoryImpl(this.authApiService, {required this.pref});

  @override
  Future<DataState<dynamic, Exception>> changePassword(String oldPassword,
      String newPassword,) async {
    final request = ChangePasswordRequest(
      oldPassword: oldPassword,
      newPassword: newPassword,
    );
    final response = await handleResponse(
          () => authApiService.changePassword(request),
    );
    return response;
  }

  @override
  Future<DataState<dynamic, Exception>> logout() async {
    return await handleResponse(() => authApiService.logout());
  }

  @override
  Future<DataState<Token, Exception>> signIn(Account account) async {
    final request = AccountRequest(
      emailOrPhoneNumber: account.emailOrPhoneNumber,
      password: account.password,
    );
    final response = await handleResponse(
          () => authApiService.signIn(request),
      fromJson: (json) => TokenDto.fromJson(json as Map<String, dynamic>),
    );
    if (response.isSuccess) {
      final TokenDto token = response.get();
      pref.setString(PrefKeys.token, token.token);
      pref.setString(PrefKeys.refreshToken, token.refreshToken ?? "");
      pref.setString(PrefKeys.role, token.role.value);
      return DataState.success(token.toToken());
    } else {
      return DataState.error(response.getError());
    }
  }

  @override
  Future<DataState<dynamic, Exception>> signUp(Account account) async {
    final request = AccountRequest(
      emailOrPhoneNumber: account.emailOrPhoneNumber,
      password: account.password,
    );
    final response = await handleResponse(
          () => authApiService.signUp(request),
    );
    return response;
  }

  @override
  Future<DataState<Token, Exception>> submitOtp(VerifyOtp verifyOtp) async {
    final request =
    VerifyOtpRequest(verifyOtp.emailOrPhoneNumber, verifyOtp.otp);
    final response = await handleResponse(
            () => authApiService.verifyOtp(request),
        fromJson: (json) => TokenDto.fromJson(json as Map<String, dynamic>));
    if (response.isSuccess) {
      final TokenDto token = response.get();
      pref.setString(PrefKeys.token, token.token);
      pref.setString(PrefKeys.refreshToken, token.refreshToken ?? "");
      pref.setString(PrefKeys.role, token.role.value);
      return DataState.success(token.toToken());
    } else {
      return DataState.error(response.getError());
    }
  }

  @override
  Future<DataState<Token, Exception>> refreshToken() async {
    final refreshToken = pref.getString(PrefKeys.refreshToken);
    if (refreshToken.isNullOrEmpty) {
      return DataState.error(Exception());
    }
    final response = await handleResponse(
            () => authApiService.refreshToken(refreshToken!),
        fromJson: (json) => TokenDto.fromJson(json as Map<String, dynamic>));
    if (response.isSuccess) {
      final TokenDto token = response.get();
      pref.setString(PrefKeys.token, token.token);
      pref.setString(PrefKeys.refreshToken, token.refreshToken ?? "");
      pref.setString(PrefKeys.role, token.role.value);
      return DataState.success(token.toToken());
    } else {
      return DataState.error(response.getError());
    }
  }

  @override
  Future<DataState<AuthConfig, Exception>> getConfig() async {
    final int counter = pref.getInt(PrefKeys.counter) ?? 0;
    AuthConfig? config = _getLocalConfig();
    if (counter % Constants.timesRefreshConfig == 0 || config == null) {
      final response = await handleResponse(
        () => authApiService.getAuthConfig(),
        fromJson: (json) => AuthConfig.fromJson(json as Map<String, dynamic>),
      );
      if (response.isSuccess) {
        config = response.get();
        _saveAuthConfig(config!);
        return DataState.success(config);
      } else {
        return DataState.success(AuthConfig(false));
      }
    } else {
      return DataState.success(config);
    }
  }

  AuthConfig? _getLocalConfig() {
    final stringConfig = pref.getString(PrefKeys.authConfig);
    if (stringConfig == null) return null;
    final jsonConfig = json.decode(stringConfig);
    return AuthConfig.fromJson(jsonConfig as Map<String, dynamic>);
  }

  void _saveAuthConfig(AuthConfig config) {
    Map<String, dynamic> map = config.toJson();
    final stringConfig = json.encode(map);
    pref.setString(PrefKeys.authConfig, stringConfig);
  }
}

File: client/mobile/Sudoo/app/src/main/java/com/sudo248/sudoo/data/repository/AuthRepositoryImpl.kt
package com.sudo248.sudoo.data.repository

import android.util.Log
import com.google.gson.Gson
import com.sudo248.base_android.core.DataState
import com.sudo248.base_android.data.api.handleResponse
import com.sudo248.base_android.ktx.state
import com.sudo248.base_android.ktx.stateOn
import com.sudo248.base_android.utils.SharedPreferenceUtils
import com.sudo248.sudoo.data.api.auth.AuthService
import com.sudo248.sudoo.data.api.auth.request.AccountRequest
import com.sudo248.sudoo.data.api.auth.request.ChangePasswordRequest
import com.sudo248.sudoo.data.api.auth.request.OtpRequest
import com.sudo248.sudoo.data.api.notification.NotificationService
import com.sudo248.sudoo.data.ktx.errorBody
import com.sudo248.sudoo.data.mapper.toToken
import com.sudo248.sudoo.domain.common.Constants
import com.sudo248.sudoo.domain.entity.auth.Account
import com.sudo248.sudoo.domain.entity.auth.AuthConfig
import com.sudo248.sudoo.domain.entity.auth.Token
import com.sudo248.sudoo.domain.repository.AuthRepository
import kotlinx.coroutines.CoroutineDispatcher
import javax.inject.Inject
import javax.inject.Singleton


/**
 * **Created by**
 *
 * @author *Sudo248*
 * @since 00:32 - 05/03/2023
 */

@Singleton
class AuthRepositoryImpl @Inject constructor(
    private val authService: AuthService,
    private val notificationService: NotificationService,
    private val ioDispatcher: CoroutineDispatcher
) : AuthRepository {
    private val gson = Gson()
    override suspend fun getAuthConfig(): DataState<AuthConfig, Exception> = stateOn(ioDispatcher) {
        val counter = SharedPreferenceUtils.getInt(Constants.Key.COUNTER, 0)
        val localConfig = getLocalConfig()
        if (counter % Constants.TIMES_REFRESH_CONFIG == 0 || localConfig == null) {
            val response = handleResponse(authService.getAuthConfig())
            if (response.isSuccess) {
                response.get().data!!.also {
                    saveLocalConfig(it)
                }
            } else {
                AuthConfig(enableOtp = false)
            }
        } else {
            localConfig
        }
    }

    override suspend fun refreshToken(): DataState<Token, Exception> = stateOn(ioDispatcher) {
        val refreshToken = SharedPreferenceUtils.withApplicationSharedPreference().getString(Constants.Key.REFRESH_TOKEN, "")
        if (refreshToken.isEmpty()) {
            throw Exception()
        } else {
            val response = handleResponse(authService.refreshToken(refreshToken))
            if (response.isSuccess) {
                val token = response.get().data!!.toToken()
                saveToken(token)
                token
            } else {
                throw response.error().errorBody()
            }
        }
    }

    override suspend fun saveToken(token: Token): DataState<Unit, Exception> = state {
        Log.d("Sudoo", "saveToken: $token")
        SharedPreferenceUtils.withApplicationSharedPreference().run {
            putString(Constants.Key.TOKEN, token.token)
            putString(Constants.Key.REFRESH_TOKEN, token.refreshToken.orEmpty())
        }
        val fcmToken = SharedPreferenceUtils.getString(Constants.Key.FCM_TOKEN, "")
        if (fcmToken.isNotEmpty()) {
            notificationService.saveToken(fcmToken)
        }
    }

    override suspend fun signIn(account: Account): DataState<Token, Exception> =
        stateOn(ioDispatcher) {
            val request = AccountRequest(
                phoneNumber = account.phoneNumber,
                password = account.password,
                provider = account.provider
            )
            val response = handleResponse(authService.signIn(request))
            if (response.isSuccess) {
                val data = response.get().data!!
                /*data.userId?.let {
                    SharedPreferenceUtils.putString(Constants.Key.USER_ID, it)
                }*/
                val token = data.toToken()
                saveToken(token)
                token
            } else {
                throw response.error().errorBody()
            }
        }

    override suspend fun signUp(account: Account): DataState<Unit, Exception> =
        stateOn(ioDispatcher) {
            val request = AccountRequest(
                phoneNumber = account.phoneNumber,
                password = account.password,
                provider = account.provider
            )
            val response = handleResponse(authService.signUp(request))
            if (response.isSuccess) {
                /*response.get().data?.let {
                    SharedPreferenceUtils.putString(Constants.Key.USER_ID, it)
                }*/
            } else {
                throw response.error().errorBody()
            }
        }

    override suspend fun generateOtp(phoneNumber: String): DataState<Unit, Exception> =
        stateOn(ioDispatcher) {
            val response = handleResponse(authService.generateOtp(phoneNumber))
            if (response.isError) {
                throw response.error().errorBody()
            }
        }

    override suspend fun verifyOtp(phoneNumber: String, otp: String): DataState<Token, Exception> =
        stateOn(ioDispatcher) {
            val request = OtpRequest(
                phoneNumber,
                otp
            )
            val response = handleResponse(authService.verifyOtp(request))
            if (response.isSuccess) {
                val token = response.get().data!!.toToken()
                saveToken(token)
                token
            } else {
                throw response.error().errorBody()
            }
        }

    override suspend fun changePassword(
        oldPassword: String,
        newPassWord: String
    ): DataState<Unit, Exception> = stateOn(ioDispatcher) {
        val request = ChangePasswordRequest(
            oldPassword,
            newPassWord
        )
        val response = handleResponse(authService.changePassword(request))
        if (response.isError) {
            throw response.error().errorBody()
        }
    }

    override suspend fun logout(): DataState<Unit, Exception> = stateOn(ioDispatcher) {
        authService.logout()
    }

    private suspend fun getLocalConfig(): AuthConfig? {
        val stringConfig = SharedPreferenceUtils.getStringAsync(Constants.Key.AUTH_CONFIG)
        if (stringConfig.isEmpty()) return null
        return gson.fromJson(stringConfig, AuthConfig::class.java)
    }

    private suspend fun saveLocalConfig(config: AuthConfig) {
        val stringConfig = gson.toJson(config)
        SharedPreferenceUtils.putStringAsync(Constants.Key.AUTH_CONFIG, stringConfig)
    }
}